src/dbnode/storage/flush.go

```diff
@@ -102,28 +105,63 @@ func (m *flushManager) Flush(
 
 	defer m.setState(flushManagerIdle)
 
-	// create flush-er
-	flushPersist, err := m.pm.StartFlushPersist()
+	namespaces, err := m.database.GetOwnedNamespaces()
 	if err != nil {
 		return err
 	}
 
-	namespaces, err := m.database.GetOwnedNamespaces()
+	// Perform three separate loops through all the namespaces so that we can
+	// emit better gauges, i.e. all the flushing for all the namespaces happens
+	// at once, then all the cold flushes, then all the snapshotting. This is
+	// also slightly better semantically because flushing should take priority
+	// over cold flushes and snapshotting.
+	//
+	// In addition, we need to make sure that for any given shard/blockStart
+	// combination, we attempt a flush and then a cold flush before a snapshot
+	// as the snapshotting process will attempt to snapshot any unflushed blocks
+	// which would be wasteful if the block is already flushable.
+	multiErr := xerrors.NewMultiError()
+	if err = m.dataWarmFlush(namespaces, tickStart, dbBootstrapStateAtTickStart); err != nil {
+		multiErr = multiErr.Add(err)
+	}
+
+	rotatedCommitlogID, err := m.commitlog.RotateLogs()
+	if err == nil {
+		if err = m.dataColdFlush(namespaces); err != nil {
+			multiErr = multiErr.Add(err)
+		}
+
+		// Snapshots must be done after a cold flush because cleanup logic just
+		// checks for a successful snapshot checkpoint file. If snapshotting
+		// went first and succeeded and then a cold flush fails, the old
+		// commitlogs would still get cleaned up. If the node then restarted,
+		// cold writes would be lost.
+		if err = m.dataSnapshot(namespaces, tickStart, rotatedCommitlogID); err != nil {
+			multiErr = multiErr.Add(err)
+		}
+	} else {
+		multiErr = multiErr.Add(fmt.Errorf("error rotating commitlog in mediator tick: %v", err))
+	}
+
+	if err = m.indexFlush(namespaces); err != nil {
+		multiErr = multiErr.Add(err)
+	}
+
+	return multiErr.FinalError()
```

Comment:
Are we sure that if there is an error here that the commit logs won't get deleted?  I think the commit log cleanup logic only checks that there is a valid snapshot, not that also cold flushes contained by those commit logs were written to disk no?

cc @richardartoul

Reply:
Ah right, that's true. We can either:
1) Return if there's an error cold flushing
2) Don't clean up commit logs unless a successful cold flush has happened too
Both run the risk of using up too much disk space if cold flushes fail, but (1) is easier to implement, so I'll do that.